## Bệnh đái tháo đường ở nữ giới Pima Indian

### 1. Định nghĩa vấn đề

* Phân tích vấn đề:
- Người da đỏ Pima tại Arizona có tỷ lệ mắc tiểu đường type 2 rất cao.
- Viện Quốc gia về Bệnh tiểu đường, Tiêu hóa và Thận (NIDDK) đã nghiên cứu liên tục từ 1965.
- Bệnh tiểu đường được chẩn đoán theo WHO: glucose huyết tương ≥ 200 mg/dL sau 2 giờ OGTT hoặc được phát hiện trong chăm sóc y tế thường quy.
- Mục tiêu: xây dựng mô hình dự báo khả năng mắc bệnh tiểu đường trong vòng 5 năm, dựa trên các yếu tố nguy cơ.

* Mô tả: 
 Dataset Pima Indians Diabetes chứa thông tin của 768 bệnh nhân nữ, ít nhất 21 tuổi, thuộc cộng đồng Pima Indian gần Phoenix, Arizona, USA. Mục tiêu là dự đoán xem một bệnh nhân có bị đái tháo đường (Class = 1) hay không (Class = 0) dựa trên 8 thuộc tính y tế.

* Dữ liệu vào:
   1. Pregnancies: Số lần mang thai, phản ánh tiền sử sinh sản có thể liên quan đến nguy cơ đái tháo đường.
   2. Glucose: Nồng độ glucose huyết tương sau 2 giờ trong bài kiểm tra dung nạp glucose (mg/dl), chỉ số quan trọng để chẩn đoán đái tháo đường.
   3. BloodPressure: Huyết áp tâm thu (mm Hg), đo áp lực máu khi tim giãn ra, liên quan đến sức khỏe tim mạch.
   4. SkinThickness: Độ dày nếp gấp da tay (mm), đại diện cho lượng mỡ dưới da, một yếu tố liên quan đến béo phì.
   5. Insulin: Mức insulin huyết thanh sau 2 giờ (mu U/ml), phản ánh khả năng sản xuất insulin của cơ thể.
   6. BMI: Chỉ số khối cơ thể (kg/m²), đo lường mức độ béo phì, một yếu tố nguy cơ lớn của đái tháo đường loại 2.
   7. DiabetesPedigreeFunction: Hàm phả hệ đái tháo đường, đánh giá nguy cơ di truyền dựa trên lịch sử gia đình.
   8. Age: Tuổi (năm), ảnh hưởng đến nguy cơ mắc bệnh do sự lão hóa.
   9. Outcome: Biến lớp (0: Không đái tháo đường, 1: Có đái tháo đường), mục tiêu chính của phân tích.
    
    - Trong đó  Hàm phả hệ đái tháo đường(DPF):
        $$
        + DPF = \frac{\sum_i K_i (88 - ADM_i) + 20}{\sum_j K_j (ALC_j - 14) + 50}
        $$
        + Được xây dựng để lượng hóa yếu tố di truyền dựa trên tiền sử bệnh tiểu đường của cha mẹ, ông bà, anh chị em, họ hàng gần.
        + DPF tăng khi số lượng thân nhân mắc bệnh tăng, tuổi mắc bệnh trẻ hơn, và mức độ quan hệ huyết thống gần.
        + DPF giảm khi thân nhân không mắc bệnh và có tuổi thọ cao.

* Lựa chọn trường hợp:
    - Tiêu chí:
        + Nữ, ≥ 21 tuổi, OGTT ban đầu không bị tiểu đường.
    - Sau 5 năm:
        + Hoặc phát hiện tiểu đường (positive case).
        + Hoặc kiểm tra lại ≥ 5 năm sau vẫn không mắc (negative case).
        + Loại bỏ trường hợp phát hiện tiểu đường trong vòng < 1 năm (do quá dễ dự báo).
    - Tổng số mẫu: 768
        + 576 mẫu cho tập huấn luyện
        + 192 mẫu cho tập dự báo (test set).

* Hướng giải quyết:
    - Sử dụng ADAP (Adaptive Learning Algorithm) – một dạng mô hình học máy dựa trên perceptron, để dự báo khả năng mắc bệnh trong vòng 5 năm.
    - Biến đầu vào gồm 8 yếu tố nguy cơ chính + nhãn phân loại (có/không mắc bệnh).

* Kết quả (từ nghiên cứu trước đó [Smith et al., 1988]):
    - ADAP đưa ra dự đoán dạng xác suất (0–1), sau đó quy đổi thành nhị phân.
    - Với cutoff 0.448, mô hình đạt:
        + Sensitivity (độ nhạy): ~76%
        + Specificity (độ đặc hiệu): ~76%
    - Đây là kết quả khá tốt trong bối cảnh dữ liệu y tế thời kỳ đó, và bộ dữ liệu trở thành benchmark nổi tiếng trong cộng đồng Machine Learning.

* Kết quả: 
    - Biến mục tiêu (Outcome): Biến nhị phân (0 hoặc 1) biểu thị trạng thái đái tháo đường
        + 0: Không bị đái tháo đường (500 mẫu).
        + 1: Bị đái tháo đường (268 mẫu).
* Thông tin bổ sung:
    - Giá trị thiếu: Dữ liệu có chứa giá trị thiếu, thường được biểu thị bằng 0 trong các cột như Glucose, BloodPressure, SkinThickness, Insulin, và BMI, cần được xử lý trong quá trình phân tích.
    - Nghiên cứu trước: Một nghiên cứu trước đây của Smith et al. (1988) sử dụng thuật toán ADAP, đạt độ nhạy (sensitivity) và độ đặc hiệu (specificity) 76% trên 192 mẫu kiểm tra sau khi huấn luyện với 576 mẫu, với ngưỡng cắt 0.448.
    - Mục tiêu phân tích: Xây dựng một mô hình dự đoán hiệu quả, so sánh với kết quả ADAP, và khám phá các yếu tố quan trọng ảnh hưởng đến nguy cơ đái tháo đường.

### 2. Chuẩn bị vấn đề

#### Mô tả vấn đề
- Bệnh tiểu đường type 2 có tỷ lệ rất cao trong cộng đồng này.
- Việc chẩn đoán chỉ dựa vào tiêu chuẩn WHO OGTT (≥ 200 mg/dL sau 2 giờ) mới cho kết luận rõ ràng, nhưng điều này chỉ xác định khi bệnh đã xuất hiện.
- Mục tiêu:
    +  Trong phần này, chúng ta sẽ thiết lập môi trường làm việc bằng cách khai báo các thư viện cần thiết, load dataset Pima Indians Diabetes, và giải thích ý nghĩa của từng cột dữ liệu để bạn dễ dàng hiểu rõ giá trị mà chúng mang lại. Đây là bước nền tảng quan trọng để tiến hành phân tích sâu hơn về nguy cơ đái tháo đường trong cộng đồng Pima Indian.
    + Nhu cầu đặt ra: dự đoán sớm nguy cơ mắc bệnh tiểu đường trong tương lai dựa trên các yếu tố nguy cơ (tuổi, BMI, insulin, huyết áp, di truyền…).
- Dữ liệu:
    -  Dataset được cung cấp dưới dạng file pima-indians-diabetes.csv, chứa 768 mẫu với 8 thuộc tính số và 1 biến lớp (Outcome). Dữ liệu có thể chứa giá trị thiếu, cần được xử lý phù hợp. 
- Thách thức:
    + Dữ liệu y tế thường có mối quan hệ phi tuyến phức tạp giữa các biến.
    + Có sự hiện diện của missing values và outliers (ví dụ insulin = 0).
    + Cần một mô hình dự đoán chính xác nhưng cũng có thể giải thích được trong ngữ cảnh y tế.

#### 2.1 Khai báo thư viện 

- **pandas** là một thư viện mạnh mẽ trong Python, được sử dụng rộng rãi cho **phân tích và xử lý dữ liệu**.
- Cung cấp cấu trúc dữ liệu chính:
  - **Series**: tương tự như một mảng 1 chiều có nhãn.
  - **DataFrame**: bảng dữ liệu 2 chiều (giống Excel hoặc SQL table).
- Các tính năng chính:
  - Đọc/ghi dữ liệu từ nhiều định dạng (CSV, Excel, SQL, JSON, ...).
  - Xử lý dữ liệu: lọc, nhóm, gộp, sắp xếp, thống kê.
  - Làm sạch dữ liệu: xử lý missing values, outliers.
  - Hỗ trợ tích hợp với NumPy, Matplotlib, Seaborn để trực quan hóa dữ liệu.

In [1]:
import pandas as pd

#### 2.2 Nạp dữ liệu

- Thực hiện việc **nạp dữ liệu gốc** từ file `pima-indians-diabetes.csv` vào Python bằng thư viện `pandas`.  
- Do file CSV không chứa tiêu đề cột (header), ta cần **tự gán tên cột** dựa trên mô tả trong tài liệu nghiên cứu gốc.  
- Kết quả: dữ liệu được lưu vào một **DataFrame** có cấu trúc rõ ràng, giúp dễ dàng phân tích, trực quan hóa và xây dựng mô hình dự báo.

In [2]:
# Đọc dữ liệu từ file CSV
data_path = 'pima-indians-diabetes.csv'  # Đường dẫn tới file dữ liệu

# Gán tên cột dựa trên mô tả trong tài liệu
column_names = ['Pregnancies', 'Glucose', 'BloodPressure', 'SkinThickness', 'Insulin', 'BMI', 'DiabetesPedigreeFunction', 'Age', 'Outcome'] # Tên các cột

# Đọc dữ liệu từ file CSV
data = pd.read_csv(data_path, header=None, names=column_names)

#### 2.3 Kiểm tra một số thông tin cơ bản

- Sử dụng các hàm cơ bản của pd đễ hỗ trợ cho chúng ta có một cái nhìn tổng quát về số liệu của file 

In [3]:
# Hiển thị 5 dòng đầu để kiểm tra
print("5 dòng đầu của dữ liệu:")
print(data.head())

5 dòng đầu của dữ liệu:
   Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   
2            8      183             64              0        0  23.3   
3            1       89             66             23       94  28.1   
4            0      137             40             35      168  43.1   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0  
2                     0.672   32        1  
3                     0.167   21        0  
4                     2.288   33        1  


In [4]:
(data==0).sum()

Pregnancies                 111
Glucose                       5
BloodPressure                35
SkinThickness               227
Insulin                     374
BMI                          11
DiabetesPedigreeFunction      0
Age                           0
Outcome                     500
dtype: int64

- **Nhận xét về dữ liệu có giá trị 0:**  
   1. Pregnancies (111 giá trị 0):
      + 0 lần mang thai là hợp lý (có những phụ nữ chưa từng mang thai).
      + Không cần xử lý coi là missing.
   2. Glucose (5 giá trị 0):
      + Về mặt y khoa, glucose máu không thể = 0.
      + Đây là giá trị bất thường → coi là dữ liệu thiếu.
   3. BloodPressure (35 giá trị 0):
      + Huyết áp tâm trương không thể = 0.
      + Các giá trị này là missing values cần xử lý
   4. SkinThickness (227 giá trị 0):
      + 0 mm nếp gấp da là không hợp lý.
      + Đây là một biến có nhiều thiếu dữ liệu (≈ 30% dữ liệu).
   5. Insulin (374 giá trị 0):
      + 0 µU/mL insulin là không thực tế.
      + Biến này có nhiều dữ liệu thiếu nhất (≈ 49%).
   6. BMI (11 giá trị 0):
      + BMI = 0 là bất hợp lý.
      + Một số ít dữ liệu cần xử lý missing.
   7. DiabetesPedigreeFunction (0 giá trị 0):
      + Không có giá trị 0, dữ liệu đầy đủ.
   8. Age (0 giá trị 0):
      + Không có giá trị 0, dữ liệu đầy đủ.
   9. Outcome (500 giá trị 0):
      + Đây là biến nhãn (class):
         + 0 = không mắc tiểu đường (500 mẫu).
         + 1 = mắc tiểu đường (268 mẫu).
      + Đây không phải dữ liệu thiếu mà là nhãn hợp lệ.

In [5]:
# Hiển thị thông tin cơ bản về dữ liệu
print(data.info())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 768 entries, 0 to 767
Data columns (total 9 columns):
 #   Column                    Non-Null Count  Dtype  
---  ------                    --------------  -----  
 0   Pregnancies               768 non-null    int64  
 1   Glucose                   768 non-null    int64  
 2   BloodPressure             768 non-null    int64  
 3   SkinThickness             768 non-null    int64  
 4   Insulin                   768 non-null    int64  
 5   BMI                       768 non-null    float64
 6   DiabetesPedigreeFunction  768 non-null    float64
 7   Age                       768 non-null    int64  
 8   Outcome                   768 non-null    int64  
dtypes: float64(2), int64(7)
memory usage: 54.1 KB
None


- Nhận xét sơ bộ
- **Cấu trúc dữ liệu:**  
   - Tổng cộng **767 dòng × 9 cột**.  
   - Tất cả các biến đều dạng số (`int` hoặc `float`). 

In [6]:
# Thống kê mô tả
print(data.describe())

       Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
count   768.000000  768.000000     768.000000     768.000000  768.000000   
mean      3.845052  120.894531      69.105469      20.536458   79.799479   
std       3.369578   31.972618      19.355807      15.952218  115.244002   
min       0.000000    0.000000       0.000000       0.000000    0.000000   
25%       1.000000   99.000000      62.000000       0.000000    0.000000   
50%       3.000000  117.000000      72.000000      23.000000   30.500000   
75%       6.000000  140.250000      80.000000      32.000000  127.250000   
max      17.000000  199.000000     122.000000      99.000000  846.000000   

              BMI  DiabetesPedigreeFunction         Age     Outcome  
count  768.000000                768.000000  768.000000  768.000000  
mean    31.992578                  0.471876   33.240885    0.348958  
std      7.884160                  0.331329   11.760232    0.476951  
min      0.000000                  

- Các thống kê mô tả chính và ý nghĩa:
    + count: số lượng mẫu dữ liệu (ở đây tất cả cột = 768).
    + mean: giá trị trung bình.
    + std: độ lệch chuẩn → mức độ phân tán quanh trung bình.
    + min, max: giá trị nhỏ nhất và lớn nhất.
    + 25%, 50%, 75% (quartiles): các mốc phân vị → dùng để đánh giá sự phân phối dữ liệu.

- Chỉ số đo lường sự biến động (Dispersion):
    + std (độ lệch chuẩn):
        + Glucose: 32 → biến thiên khá lớn, thể hiện sự khác biệt giữa người khỏe mạnh và người có nguy cơ cao.
        + Insulin: 115, lớn hơn cả mean (~80) → chứng tỏ dữ liệu rất phân tán và nhiều ngoại lệ.
        + BMI: 7.88 → tương đối ổn định, dao động vừa phải.
        + Age: 11.8 → đa dạng độ tuổi, từ 21 đến 81 tuổi.
- Khoảng (range = max - min):
    + Pregnancies: 0–17, cho thấy có vài trường hợp đặc biệt mang thai rất nhiều lần.
    + Insulin: 0–846, cực kỳ phân tán → cần chuẩn hóa hoặc log-transform khi phân tích.
    + SkinThickness: 0–99, cũng có nhiều giá trị bất hợp lý.

In [7]:
# Phân bố lớp
print("\nPhân bố biến lớp (Outcome):")
print(data['Outcome'].value_counts())


Phân bố biến lớp (Outcome):
Outcome
0    500
1    268
Name: count, dtype: int64


- Tỉ lệ mắc ~35% và không mắc ~65%